### PR TITLE
implement retry logic with twilio conferences api

### PR DESCRIPTION
**Pull request by jbraswell**
Sometime in August 2023, Twilio introduced a change in API Behavior. The conferences API now seems to be eventually consistent. Sometimes we get the conference back on the first try, and other times it takes a few tries. Retrying every half second seems to get the job done after no more than about 4 retries in the worst case. We try up to 10 times just to be safe.

Without this logic, there's an intermittent bug where callers never get connected to volunteers, they just listen to the hold music forever. We probably need to ship this ASAP with some sort of announcement. With the debug logging, I can see that the SEZF encountered this eventual consistency bug 20 times on Friday.

NOTE: Someone should run this code once to make sure it works. I ported it from a 4.2.1 install, and the code (particularly how logging and twilio client retrieval works) has changed since then.

### CODE REVIEW

File: app/Http/Controllers/HelplineController.php [hunk 1]
```diff
@@ -234,8 +234,28 @@ public function dial(Request $request)
...
+        // try, and other times it takes a few tries. Retrying every half second seems to get the
+        // job done after no more than about 4 retries in the worst case. We try up to 10 times just
+        // to be safe.
+        $conferences = array();
+        for ($i = 0; $i < 10; $i++) {
+            $conferences = $this->twilio->client()
+                ->conferences
+                ->read(array("friendlyName" => $request->get('FriendlyName')));
+
+            if ($i > 0) {
```
Comment (cjwillis48) [inline app/Http/Controllers/HelplineController.php:248]: You could start the loop at 1, and put the log just before the sleep to remove a conditional

Comment (jbraswell) [reply app/Http/Controllers/HelplineController.php:248]: The intent is to print a log line every time it retries, not skipping the last try.

If I did what you're suggesting, wouldn't the last (and successful) retry not print a log line?

Not sure what starting the loop at 1 accomplishes - all of the conference read requests are happening inside the loop, so the first try is still the first iteration.

Perhaps I'm misunderstanding what you're suggesting?

Comment (cjwillis48) [reply app/Http/Controllers/HelplineController.php:248]: It's pedantic, it's not really a change I'd require to approve the PR, just a preference thing. 
 
We are taking two slightly different logging approaches really. For you i represents the retries number while mine is the attempt number. Your way is logging that you made a retry and my way logs that no conferences were found in that attempt.

Mine would be more like 

"Conferences eventual consistency issue occurred in attempt 1"

**Comment by dgershman** [comment]
I wonder if I upgrading the Twilio Client would make a difference too.  @cjwillis48 are you able to test this out, if I backport it to 4.2.10 branch?

**Comment by cjwillis48** [comment]
> I wonder if I upgrading the Twilio Client would make a difference too. @cjwillis48 are you able to test this out, if I backport it to 4.2.10 branch?

We're on 4.2.1 still on our server. I could probably upgrade it to 4.2.10 if someone walks me through it once or just watches me to make sure I don't screw it up. But yeah I can absolutely test

**Comment by pjaudiomv** [comment]
> I wonder if I upgrading the Twilio Client would make a difference too. @cjwillis48 are you able to test this out, if I backport it to 4.2.10 branch?

I was curious about this too like maybe newer lib has retry logic in it too or doesn't need it

**Comment by jbraswell** [comment]
I tried upgrading the twilio client and it didn't help.

This is a timing issue.

If you want to backport to 4.2.x, I have a thread in Slack with that logic.

**Comment by dgershman** [comment]
You would be willing to add that logic to this branch? https://github.com/bmlt-enabled/yap/tree/hotfix/4.2.11

From: jbraswell ***@***.***>
Date: Sunday, August 27, 2023 at 11:43 AM
To: bmlt-enabled/yap ***@***.***>
Cc: Danny Gershman ***@***.***>, Review requested ***@***.***>
Subject: Re: [bmlt-enabled/yap] implement retry logic with twilio conferences api (PR #827)

I tried upgrading the twilio client and it didn't help.

This is a timing issue.

If you want to backport to 4.2.x, I have a thread in Slack with that logic.

**Comment by jbraswell** [comment]
Not today. I could in a few days. I was saying you could just copy/paste from slack if you wanted to backport.

**Comment by dgershman** [comment]
Is it this?

```php
$conferences = [];
for ($i = 0; $i < 7; $i++) {
    $conferences = $twilioClient->conferences->read(array ("friendlyName" => $_REQUEST['FriendlyName'] ));
    if (count($conferences)) {
        break;
    }
    sleep(0.5);
}
```

From: jbraswell ***@***.***>
Date: Sunday, August 27, 2023 at 12:04 PM
To: bmlt-enabled/yap ***@***.***>
Cc: Danny Gershman ***@***.***>, Review requested ***@***.***>
Subject: Re: [bmlt-enabled/yap] implement retry logic with twilio conferences api (PR #827)

Not today. I could in a few days. I was saying you could just copy/paste from slack if you wanted to backport.

**Comment by cjwillis48** [comment]
@dgershman I can add the code to that branch later if you want. Probably around. 9 EST

**Comment by dgershman** [comment]
I made the change and released 4.2.11-beta1 for you test again.

**Review by dgershman — APPROVED** [review]

**Comment by dgershman** [comment]
lgtm

**Pull request merged**